### PR TITLE
WIP: Offline support

### PR DESCRIPTION
**Pull request by anandksalien**
**Service Worker**
Added a Service worker which assimilates the code generated by the offline-plugin to add some dynamic caching mechanism (to cache word templates , images , font files ) etc.

**Disabled options**
Disabled Certain options in editor when the user goes offline.Disabled options are sharing a document , creating a copy, creating a copy with a different template , exports related to ePub , LaTex , ODT, JATS , and movement between different document styles.

When user opens a document the requests related to images and fonts and the docx templates are cached by the service worker and put into a separate cache. This uses a cache-first approach. So when the user goes offline the images and everything can be served by the service worker as they are cached already.
**NOTE** When the user opens developer tool window in chrome and uses the disable cache option the print PDF will not work properly because the requests made by Vivliostyle from an Iframe are not intercepted by the Service worker so the PDF will contain broken Images. and also when the user goes offline editor's currentView is used to get the updated doc for exports and printing.

**Indexed DB** 
A general structure for indexed DB is created with some common methods to create Objectstore , readdata from objectStore etc. The document overview page and the contacts page's JSON response are cached and is stored in the respective Object stores.

**Login cookie expiration**
In case the login cookie expires when the user was offline , the document will downloaded as a fidus export file format and user would be moved onto the login page.

**Bibliography**
When user tries to add a bibliography to his/her own collection , i.e., from the bibliography page , it will show the generic offline message from the post handler , and inside the document since the bibliography is stored inside the editor object , it is enabled for now, so the users can register new sources.

**Normal Merge**
In case of normal merge without any conflicts the diff window will not be shown and the document is merged automatically .

**Change in Recreate transform**
Now recreate transform uses Word diffing instead of character diffing.

**Diff Merge**
In case if the changes are conflicting , the user who came back online will see a merge window with multiple editors allowing him/her to restructure the document according to their needs. Also the main editor in this case would be aligned with the latest online version of the document , and when the user who is merging the document clicks on merge complete , the merged document is put into the main editor by rebasing the steps.
Some images for merge instructions:
screenshot1
screenshot2

**Tests that are added**
New tests have been added to test the new functionalities namely the login cookie expiration , working of SW , working on indexedDB , and a simple test to see if conflicting changes do lead to opening of a merge window.

**Improvements that can be done**
These are list of improvements that can be done to the current implementation of diff based merging.
1.  Since the plugins are directly bound to the main view , if we have plugins that are more generic in nature , we can use all the plugins in the diff editor. One useful example for this would be the fact that since the image/table plugin isn't enabled in the diff editor , sometimes the table may not be exactly rectangular, or the users can't edit the attributes of the images in diff editor.
2. Having a merge restart button , in case the users want to start merging from the start and also the ability to highlight the changes that are conflicting in nature and also the ability to preview the changes before they're applied. Also the ability to handle page refreshes (Now if a page is refreshed the offline content is lost )
3. Having separate scroll bars for each editors - this will help users look at the same paragraphs in different editors (Note that editors might have different content so the paragraphs may not be aligned)
4. Try to implement a merge like in VS code - where we have only one merge window and all the conflicts are present - This would mean that somehow we re-insert the deleted content though which might be bit tricky.
5. Displaying comments/instructions/Toolbar/Headerbar in the merge window.
6. Letting other user's know that merge is happening to the document.
7. When a footnote is added by online user , and if the offline user added text in next paragraph while trying to merge with the diff window , the offline user's change cannot be accepted automatically. - This happens because of the step we get from the recreate transform library , and this causes mapping issues with prosemirror. (step when mapped becomes null)
8. Handling of tracked changes - assume that there was an user who had write-track changes access right , and owner of the document went offline , and during this time period assume that the user with write tracked access made changes to document , when the owner comes back online (assume that track changes option is disabled. because when it is enabled whatever the user manually types in the middle editor will be tracked automatically. also note that even if track changes is enabled in middle editor accepting a merge change isn't tracked.) and tries to merge the document , if for some reason the accept changes didn't work , the user would have to copy paste the content and while doing so he'll lose the tracked changes.  To overcome this we should add accept with track changes options and also an option for user to say mark certain insertions/deletions as tracked changes.
9. Splitting of steps that delete and insert in the same step into multiple steps , that just inserts and deletes. Improve error handling for this. In case if the step couldn't be split we just return the original transaction. Should look for different solutions for this.
10. When we use the recreate transform to retrace the lost steps , when the user inserts a footnote the step we get from the library says delete the paragraph/text node after the position of insertion of footnote and replace it with a footnote node and the text node. This leads to some loss of marks issue when we merge the documents together. the marks that would be most important include the mark related to track changes. This situation would come into play when we have a normal merge without conflict , and when we try to merge the merged document with the latest version of the document when the user completes resolving the diff merge.

### CODE REVIEW

**Comment by johanneswilm** [comment]
Hey @anandksalien . This looks really good. Can you check the failed PR though? I think there are some linting issues and maybe some other things. How close is this to being ready to be merged?

**Comment by johanneswilm** [comment]
@anandksalien The difference between `post()` and `postJson()` is that `postJson()` performs `post()` and then processes the return json. So whenever no json is being returned from the server, we just use `post()`. I think you can move your online check to the `post()` function, right?

File: fiduswriter/base/root_urls.py [hunk 1]
```diff
@@ -48,6 +48,14 @@
         'path': 'sw.js'
     }),
 
+    url('^sw2.js$', static_serve, {
```
Comment (johanneswilm) [inline fiduswriter/base/root_urls.py]: @anandksalien Are we using both sw2.js and sw.js or is this replacing sw.js?

Comment (anandksalien) [reply fiduswriter/base/root_urls.py]: We need the transpiled version of sw.js and we import it in sw2.js. so we would be needing both sw.js and sw2.js.(sw.js needs to be accessible via an URL endpoint)

**Comment by anandksalien** [comment]
> anandksalien

Yeah will do

**Comment by anandksalien** [comment]
Hey @johanneswilm do we have to cache the document data in the indexed DB as well ? (could be useful when the user goes offline and refreshes the page ?)

**Comment by johanneswilm** [comment]
@anandksalien Yes, we should. But maybe we should rethink how the user obtains the document in the first place and also which documents we cache. Some things to think about:

* Previously we sent the document through the ws connection because that ensured that the user would get the latest version of the document. But our system should now be so advanced that if the user gets a document version that is slightly older, the user's browser can simply ask for the remaining changes through the ws connection.

* The ws connection has to be fast and blocking and will always have to be served from the same process on the server. Calls to other views (like get_document) can be served by other process.

* The document data should be the same as what is downloaded as the document list data + the data it gets from https://github.com/fiduswriter/fiduswriter/blob/develop/fiduswriter/document/views.py#L36 . So when you cache the document list, one could for example already cache the extra data for the latest X documents (by change/access date) and with some changes for how document data is loaded, that should also make it possible to load those in the editor.

**Comment by anandksalien** [comment]
@johanneswilm  Do we allow users to use the image that they've uploaded already when they're offline ? Because if so then we need to handle the scenario where in the image got deleted in the server and the document and when the user comes online the image is lost ? (with current implementation the users can't add images to document( when offline) they've uploaded already as the button in toolbar is disabled)

**Comment by johanneswilm** [comment]
@anandksalien I think we can disable uploading new pictures altogether while the user is offline at least for the first working version. But we still have to deal with the issue what happens if an image has been deleted on the server while the user was offline if the user is using that image. You should investigate whether it is possible to store an image with a specific id, so that if say Image 654 has been deleted while the user was offline, but he had copied and pasted that image from one place in the document to another so that it is in the document again when he comes back online - can we upload and save that image as image 654 again? Or does it need to get a new id because that field is auto updating?

**Comment by johanneswilm** [comment]
>  (with current implementation the users can't add images to document( when offline) they've uploaded already as the button in toolbar is disabled)

Yes, but what if the user uses copy and paste within the same document? Say the user copies picture 645 from 1st to 3rd section in the document. Another user deletes the image from section 1 in the meantime. When the two edits are merged together, it will delete the image from section 1 (both users deleted it there), but it will then add it to section 3. So unless you also disable copy and paste, there will be ways for users to make an image that previously was deleted reappear.

You probably also need to think about and test what happens if users paste content, including images from other documents.

File: fiduswriter/base/static/js/modules/app/index.js [hunk 2]
```diff
@@ -1,6 +1,5 @@
-import OfflinePluginRuntime from 'offline-plugin/runtime'
 import {CSL} from 'citeproc-plus'
-
+import OfflinePluginRuntime from 'offline-plugin/runtime' //eslint-disable-line
```
Comment (johanneswilm) [inline fiduswriter/base/static/js/modules/app/index.js]: @anandksalien Why do we need/want to import OfflinePluginRuntime if we then don't use it anywhere in the file?

Comment (anandksalien) [reply fiduswriter/base/static/js/modules/app/index.js]: Hey @johanneswilm . I checked this out. We need the import, only then the sw.js is generated and put into static-transpile folder. If this is not imported the sw.js is not generated. Please refer this link for more details [https://goo.gl/YwewYp]( https://goo.gl/YwewYp)

Comment (johanneswilm) [reply fiduswriter/base/static/js/modules/app/index.js]: Hey @anandksalien on that link they also call the install method. But in our case we remove the install call.

Comment (anandksalien) [reply fiduswriter/base/static/js/modules/app/index.js]: @johanneswilm Yeah I noticed it. But for some reason when I try and remove that import it throws this error while collecting static and doesn't generate the sw.js that is needed.
image
This was the reason why i kept it for now.Don't know if it's really an issue with the library.

**Pull request merged**